### PR TITLE
CLIPA Torch Train

### PR DESCRIPTION
**Pull request by zw615**
Add code for syntax masking, and text pos_embed resizing. More color jitter and gray scale options added. Also include some example model configs and runing scripts

### CODE REVIEW

**Comment by mehdidc** [comment]
Thanks a lot @zw615 for the PRs! I think we can already start to reproduce some of the experiments. 
Just to be sure, syntax masking only works for english right? or does it automatically detect language?
maybe for the non-supported cases, it would make sense to support block masking, or random masking,
that were mentioned in the paper.

**Comment by zw615** [comment]
For syntax masking, we opt for the popular Natural Language Toolkit (NLTK) for tokenization and tagging. Personally I am not that familiar with this package. To the best of my knowledge, it does not support multilingual tagging. But the general idea of syntax masking should be easily applicable with other languages when equipped with a proper NLP library.

As for block masking or random masking, I can merge and push them into this pr if you want.

**Comment by JeniaJitsev** [comment]
> As for block masking or random masking, I can merge and push them into this pr if you want.
@zw615 merging block/random masking for this pr sounds good

**Comment by zw615** [comment]
> > As for block masking or random masking, I can merge and push them into this pr if you want.
> > @zw615 merging block/random masking for this pr sounds good

Sorry, kind of busy this week. Will pick it up after a week or so.

**Comment by mehdidc** [comment]
Thanks @zw615 would be cool, could you also please mention `nltk` as a requirement, here since it is optional, I think would make sense to  just mention it in a CLIPA section in the README

**Comment by zw615** [comment]
Block masking and random masking for text are pushed. Let me know if you have any question.


> Thanks @zw615 would be cool, could you also please mention `nltk` as a requirement, here since it is optional, I think would make sense to just mention it in a CLIPA section in the README

As for mentioning `nltk` as a requirement, are you saying that I should open a new CLIPA section in the README file? Should I write some intro about CLIPA? Or just add `nltk` in the `requirements-training.txt`?

**Comment by zw615** [comment]
Hi, just finished the stuff on my hand. I will be available for the next couple of days. Let me know if you have any more question.

**Comment by mehdidc** [comment]
@zw615 Sorry for the delay and thanks a lot for the update.
Many thanks again for the PR, I have been experimenting a bit with B-32, and everything works fine.

> As for mentioning nltk as a requirement, are you saying that I should open a new CLIPA section in the README file? Should I write some intro about CLIPA? Or just add nltk in the requirements-training.txt?
.
I meant writing an intro about CLIPA, yes, because adding `nltk` to requirements would be too constraining I think, as it is optional. Additionally, as CLIPA training is multi-stage, I think it would be nice to have a more detailed doc on how to train a CLIPA model (e.g. the L/14 one), this could be in docs/ as a markdown file, like https://github.com/mlfoundations/open_clip/blob/main/docs/datacomp_models.md.


Few other remarks/questions:

- I got issues with `nltk` in offline mode, which is a usual setup in supercomputers, running with `nlk.download` freezes the job. I had to first download nltk data manually in a folder (nltk.download('punkt', download_dir="cache") and nltk.download('averaged_perceptron_tagger', download_dir="cache"), then use the env variable `export NLTK_DATA=cache` in the training script, and commenting out the `nltk.download` lines in  `src/open_clip/tokenizer.py`, to avoid downloading them. Could you please document this case?
- Are the provided scripts the exact hyper-parameters that would be used to reproduce H/14 results ?  or just an example ? I think it would be nice to have the exact scripts 
-  There is a folder `[script_examples](https://github.com/mlfoundations/open_clip/tree/main/docs/script_examples)` where you can put the scripts, e.g. in a `clipa` subfolder
- Would be nice to support the existing L/14 and H/14 checkpoints (https://github.com/UCSC-VLAA/CLIPA) from the paper directly in openclip, e.g. supported in `create_model_and_transforms`
-
Sorry for the long list :),  I can also take care of the stuff I mentioned if you prefer/can't do it right now, e.g. the docs.

**Comment by zw615** [comment]
Hi @mehdidc, sorry for the late reply. Sure, I can write a markdown file like `datacomp_models.md`. As for the other remarks,

- I will document this nltk case in the markdown file.
- No, they are not. this script only pretrain the model for 7 epochs for demonstration. To compete with SOTA models, we pretrain the model with 12.8B on LAION-2B or DataComp-1B dataset, and the finetuning schedule is also prolonged. The original models are trained with JAX, since it scales better with TPU. I will need to check our jax configs and scripts, and convert them into pytorch scripts. As it is infeasible for us to retrain those models again with pytorch, I cannot verify its correctness 100%. But I think we still keep the original log files for the H/14 training. They should be helpful when reproducing the H/14 results. Also, you should be able to see the config files we used in `clipa_jax/configs/model_h`.
- Cool! I will update this pr and put the scripts in a `clipa` subfolder.
- Yes, it is possible to support the existing L/14 and H/14 checkpoints. But we trained those large models with our JAX implementation. Those pytorch checkpoints are converted from JAX afterwards. Because of some subtle implementation differences (e.g. tokenizer), I have put them in another PR (https://github.com/mlfoundations/open_clip/pull/579). Check this PR to decide if you want to merge the changes into open_clip.

I will gradually finish this list as it is now approaching some submission deadlines. Thanks for your time and patience!

**Comment by mehdidc** [comment]
Alright looks very good, many thanks @zw615 and all the best with your submission!

**Comment by zw615** [comment]
Hi @mehdidc , a new commit was just pushed, and it contains a markdown file `clipa.md`, and some example scripts under `script_examples`, with introductions about installation, training, testing, and some trained weights of CLIP. I have also include scripts trying to reproduce our best CLIPA-v2 H/14 experiment, but their correctness has not been verified. 

The existing CLIPA-v1 checkpoints are already open_clip compatible, i.e., supported in `create_model_and_transforms`. As for the CLIPA-v2 checkpoints, check the other PR (https://github.com/mlfoundations/open_clip/pull/579) to decide if you want to merge them first.

**Comment by rom1504** [comment]
Looks great.

I think it's pretty much ready for merging.

One thing left though: the CI is failing. Could you fix that please?
